### DISCUSSION TITLE
WASM, Polymorphism and JSON-Serialization

### DISCUSSION THREAD

**Discussion by TopperDEL** [unoplatform/uno#16925] [Q&A] [answered]
I have a class structure like this: IInterface => BaseClass => SubClass.

Now I get a complex object via an API, that uses IInterface. To support the creation of a concrete type for that interface, I've added a TypeDiscriminator-property and added a JsonConvert.

This all works fine everywhere except on WASM. On WASM, `System.Text.Json` is not able to deserialize my IInterface-Type using that Converter and the type discriminator. I've searched for solutions for two days now. First I suspected trimming and AOT, but that is not the case. Even if I disable trimming or explicitly use the SubClass, it does not work. It seems that `System.Text.Json` behaves different on WASM then on e.g. Windows Desktop.

Any ideas what I could do? Maybe I'm searching for the wrong keywords or it is a problem with trimming/AOT nevertheless?

Here some code:

```
public interface IInterface
{
     string Name { get; set; }
}

public class BaseClass : IInterface
{
    public string Name { get; set; }
    public string TypeDiscriminator { get; set; }
}

public class SubClass : BaseClass
{
    public string SubProperty { get; set; }
}
```

Then I have a JsonSerializerOptions like this:
```
public static class DefaultJsonSerializerOptions
{
     public static System.Text.Json.JsonSerializerOptions Options { get; private set; }
     static DefaultJsonSerializerOptions()
     {
         Options = new System.Text.Json.JsonSerializerOptions
         {
             WriteIndented = true,
             IncludeFields = true,
             Converters =
             {
                     new TypeDiscriminatorConverter<IInterface>()
             }
         };
     }
 }
```

And this TypeDiscriminatorConverter:
```
public class TypeDiscriminatorConverter<T> : JsonConverter<T> where T : ITypeDiscriminator
{
     private readonly IEnumerable<Type> _types;

     public TypeDiscriminatorConverter()
     {
         var type = typeof(T);
         _types = AppDomain.CurrentDomain.GetAssemblies()
             .Where(t => t.GetName().Name != "uplink.NET")
             .SelectMany(s => s.GetTypes())
             .Where(p => type.IsAssignableFrom(p) && p.IsClass && !p.IsAbstract)
             .ToList();
     }

     public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
     {
         if (reader.TokenType != JsonTokenType.StartObject)
         {
             throw new JsonException();
         }

         try
         {
             using (var jsonDocument = JsonDocument.ParseValue(ref reader))
             {
                 if (!jsonDocument.RootElement.TryGetProperty(nameof(ITypeDiscriminator.TypeDiscriminator), out var typeProperty))
                 {
                     throw new JsonException();
                 }

                 var type = _types.FirstOrDefault(x => x.Name == typeProperty.GetString());
                 if (type == null)
                 {
                     throw new JsonException();
                 }

                 var jsonObject = jsonDocument.RootElement.GetRawText();
                 var result = (T)JsonSerializer.Deserialize(jsonObject, type, options);

                 return result;
             }
         }
         catch(Exception ex)
         {
             return default;
         }
     }

     public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
     {
         JsonSerializer.Serialize(writer, (object)value, options);
     }
 }
```

Now I do something like this:
```
IInterface myObj = new SubClass{ SubProperty = "1234" };
var json = JsonSerializer.Serialize(myObj, DefaultJsonSerializerOptions.Options);
var deserMyObj = JsonSerializer.Deserialize<IInterface>(json, DefaultJsonSerializerOptions.Options);
```

This works on windows but fails on WASM, and I don't know why.

**Comment by jeromelaban** [comment]
Thanks for the question. Can you post the error you're getting? Also, make sure to take a look at the [linker documentation](https://platform.uno/docs/articles/features/using-il-linker-webassembly.html).

**Reply by TopperDEL** [reply]
Thanks @jeromelaban for you quick reply!
The error is mostly just an empty field where the JSON-string has data in it - or sometimes an exception stating that it does not know how to deserialize. It depends on how I test - I did multiple versions and tried different stuff, so I currently cannot really give one exception.

I also read the linke documentation you referenced multiple times. What I'm asking myself currently: the `<WasmShellILLinkerEnabled>false</WasmShellILLinkerEnabled>` - do I need to set it only on the WASM-project? Or also for one of my class library that are involved in my deserialization (e.g. the Contracts-project containing my Interface and/or the BaseClass)? This is something I may not have tried yet.

I'm also checking the `_types'-list during debug and release and they are different - which leads to a trimming issue then. What wonders myself again: the type im testing here is included in both versions (release and debug), so I would assume that everything is in place.

Hm, I will check again with the linker and the `WasmShellILLinkerEnabled`property. If you have ideas left, let me know! Thank you in advance!

**Reply by TopperDEL** [reply, marked as answer]
So, I finally found my error: the linking/trimming is the problem, indeed. But I was focusing on the wrong project. I thought the issue is within the server-project and/or that disabling the linker there would help. But I needed to adjust the linker-process withing my WASM-app that gets served by the server-project. Setting `<WasmShellILLinkerEnabled>false</WasmShellILLinkerEnabled>` there resolves my issue, so adding a Linkerconfig-xml there should be my target.

Sorry for the confustion. I mostly confused myself. For days. 🙈
